app/services/ipam/external_ipam_service.rb

```diff
@@ -0,0 +1,115 @@
+module IPAM
+  class ExternalIpamService < Base
+    delegate :external_ipam_proxy, :to => :subnet
+
+    @@ipam_reservations = nil
+    @@cleaner = nil
+    # Max times to retry for a new IP in the case of a race condition
+    # when multiple hosts are being provisioned simultaneously
+    # TODO: Make this configurable?
+    MAX_RETRIES = 5
+    # The delay to wait in seconds before clearing the in memory cache
+    CLEAR_CACHE_DELAY = 180
```

Comment:
I am afraid this approach will not work. Foreman uses multiple processes, this means that every process will get its own cache = race conditions.

What I was initially suggesting was to take the similar approach which our DHCP smart proxy module does in `unused_ip` call: it "remembers" list of pre-allocated IP and if there is another call requesting new suggested IP it will not use it. There is a helper class FreeIps in the DHCP common module which you can use in your Smart Proxy module (just make dhcp_common as a dependency of your module and you can use it). Since Smart Proxy is a multi-threaded app (single process), there is a mutex that protects concurrent access and this works fine. There is a background thread which cleans up allocated IP addresses every X minutes (configurable interval).

It should be pretty easy to implement this on the Smart Proxy, here in Foreman core we do not want any of this code - here it should be just simple HTTP(s) call. Apologies if this was not clear from the beginning, my fault.

Reply:
Thanks @lzap - I will take a look at the DHCP implementation.